README.md
# React CSS Transition Replace

A [React](http://facebook.github.io/react/) component to animate replacing one element with another.

While [`ReactCSSTransitionGroup`](https://facebook.github.io/react/docs/animation.html) does a great job
of animating changes to a list of components and can even be used to animate the replacement of one item
with another, proper handling of the container height in the latter case is not built in. This component
is designed to do exactly that with an API closely following that of `ReactCSSTransitionGroup`.

Using `react-css-transition-replace` provides two distinct benefits:

- It automatically handles the positioning of the animated components, and
- _allows changes in the height of container to be handled and animated with ease when
  various content heights differ, even when absolute positioning is used._

Animations are fully configurable with CSS, including having the entering component wait to enter until
the leaving component's animation completes. Following suit with the
[React.js API](https://facebook.github.io/react/docs/animation.html) the one caveat is
that the transition duration must be specified in JavaScript as well as CSS.

[Live Examples](https://marnusw.github.io/react-css-transition-replace) |
[Change Log](/CHANGELOG.md) |
[Upgrade Guide](/UPGRADE_GUIDE.md)

## Installation

Install via `npm`:

```
npm install --save react-css-transition-replace
```

## Important Note

All functional child components must be wrapped in `forwardRef` to work correctly.

## Usage

A `ReactCSSTransitionReplace` component can only have a single child. Other than that, the basic usage
follows the exact same API as `ReactCSSTransitionGroup`, with support for `transitionEnter`, `transitionLeave`
and `transitionAppear`. When the `key` of the child component changes, the previous component is animated out
and the new component animated in. During this process:

- All leaving components continue to be rendered; if the animation is slow there may be multiple components
  in the process of leaving.
- The entering component is positioned on top of the leaving component(s) with `absolute` positioning.
- The height of the container is set to that of the leaving component, and then immediately to that of the
  entering component. If the `transitionName` is a `String` the `{animation-name}-height` class name is applied
  to it, and if `transitionName` is an `Object` the `transitionName.height` class will be used if present.
- The leaving component will be passed an `isLeaving` prop while transitioning out.

This provides many possibilities for animating the replacement as illustrated in the examples below.

Additionally, the boolean property `changeWidth` can be used to animate changing the width of the component.
This change will happen at the same time as changing the height. Animating this change should be done using
the same class name that is used for animating the change in height.

It is also possible to remove the child component (i.e. leave `ReactCSSTransitionReplace` with no children)
which will animate the `height` going to zero along with the `leave` transition. Similarly, a single child
can be added to an empty `ReactCSSTransitionReplace`, triggering the inverse animation.

By default a `span` is rendered as a wrapper of the child components. Each child is also wrapped in a `span`
used in the positioning of the actual rendered child. These can be overridden with the `component` and
`childComponent` props respectively.

### Cross-fading two components

The `ReactCSSTransitionReplace` component is used exactly like its `ReactCSSTransitionGroup` counterpart:

```javascript
import ReactCSSTransitionReplace from 'react-css-transition-replace';

render() {
  return (
    <ReactCSSTransitionReplace transitionName="cross-fade"
                               transitionEnterTimeout={1000} transitionLeaveTimeout={1000}>
      <SomeComponent key="uniqueValue"/>
    </ReactCSSTransitionReplace>
  );
}
```

To realize cross-fading of two components all that remains is to define the enter and leave opacity
transitions in the associated CSS classes:

```css
.cross-fade-leave {
  opacity: 1;
}
.cross-fade-leave.cross-fade-leave-active {
  opacity: 0;
  transition: opacity 1s ease-in;
}

.cross-fade-enter {
  opacity: 0;
}
.cross-fade-enter.cross-fade-enter-active {
  opacity: 1;
  transition: opacity 1s ease-in;
}

.cross-fade-height {
  transition: height 0.5s ease-in-out;
}
```

Note the additional `.cross-fade-height` class. This indicates how the container height is to be
animated if the heights of the entering and leaving components are not the same. You can see this
in action [here](http://marnusw.github.io/react-css-transition-replace#cross-fade).

### Fade out, then fade in

To fade a component out and wait for its transition to complete before fading in the next, simply
add a delay to the `enter` transition.

```css
.fade-wait-leave {
  opacity: 1;
}
.fade-wait-leave.fade-wait-leave-active {
  opacity: 0;
  transition: opacity 0.4s ease-in;
}

.fade-wait-enter {
  opacity: 0;
}
.fade-wait-enter.fade-wait-enter-active {
  opacity: 1;
  /* Delay the enter animation until the leave completes */
  transition: opacity 0.4s ease-in 0.6s;
}

.fade-wait-height {
  transition: height 0.6s ease-in-out;
}
```

_Note:_ The `transitionEnterTimeout` specified in the JS must be long enough to allow for the delay and
the duration of the transition. In this case:

```javascript
<ReactCSSTransitionReplace transitionName="fade-wait"
                           transitionEnterTimeout={1000} transitionLeaveTimeout={400}>
```

See the live example [here](http://marnusw.github.io/react-css-transition-replace#fade-wait).

### React-Router v4

Animated transitions of react-router v4 routes is supported with two caveats shown in the example below:

1. The current `location` must be applied to the `Switch` to force it to maintain the previous matched route on
   the leaving component.
2. If the `Switch` might render `null`, i.e. there is no catch-all `"*"` route, the `Switch` must be wrapped in a
   `div` or similar for the leave transition to work; if not the previous component will disappear instantaneously
   when there is no match.

```javascript
<Router>
  <div className="router-example">
    <ul>
      <li>
        <Link to="/">Home</Link>
      </li>
      <li>
        <Link to="/one">One</Link>
      </li>
      <li>
        <Link to="/two">Two</Link>
      </li>
      <li>
        <Link to="/three">Three (no match)</Link>
      </li>
    </ul>
    <Route
      render={({ location }) => (
        <ReactCSSTransitionReplace
          transitionName="fade"
          transitionEnterTimeout={500}
          transitionLeaveTimeout={500}
        >
          <div key={location.pathname}>
            <Switch location={location}>
              <Route path="/" exact component={Home} />
              <Route path="/one" component={One} />
              <Route path="/two" component={Two} />
            </Switch>
          </div>
        </ReactCSSTransitionReplace>
      )}
    />
  </div>
</Router>
```

See the live example [here](http://marnusw.github.io/react-css-transition-replace#react-router-v4).

### Hardware acceleration for smoother transitions

For smoother transitions hardware acceleration, which is achieved by using translate3d instead of the 2D
translations, should be used whenever possible. For example, to realize a mobile app transition between
pages one might use:

```css
.page-enter,
.page-leave {
  position: absolute;
  -webkit-transition: transform 250ms ease-in-out, opacity 250ms ease-in-out;
  transition: transform 250ms ease-in-out, opacity 250ms ease-in-out;
}

.page-enter {
  left: 100vw;
}

.page-enter.page-enter-active {
  -webkit-transform: translate3d(-100vw, 0, 0);
  transform: translate3d(-100vw, 0, 0);
}

.page-leave {
  left: 0;
}

.page-leave.page-leave-active {
  -webkit-transform: translate3d(-100vw, 0, 0);
  transform: translate3d(-100vw, 0, 0);
}
```

```javascript
<ReactCSSTransitionReplace
  transitionName="page"
  transitionEnterTimeout={250}
  transitionLeaveTimeout={250}
>
  <div key="page01">My page 01 content</div>
</ReactCSSTransitionReplace>
```

## Tips

1.  In general animating `block` or `inline-block` level elements is more stable that `inline` elements. If the
    height changes in random ways ensure that there isn't a `span` or other inline element used as the outer
    element of the components being animated.
2.  The `overflow` of the container is set to `'hidden'` automatically, which changes the behaviour of
    [collapsing margins](https://css-tricks.com/what-you-should-know-about-collapsing-margins/) from the default
    `'visible'`. This may cause a glitch in the height at the start or end of animations. To avoid this you can:
    - Keep the overflow hidden permanently with custom styles/classes if that will not cause undesired side-effects.
    - Only use
      [Single-direction margin declarations](http://csswizardry.com/2012/06/single-direction-margin-declarations/)
      to avoid collapsing margins overall.
    - Turn this feature off by setting the `overflowHidden={false}` prop when hidden overflow is not needed,
      for example when transitions are in place and content is of the same height.
3.  If the `.*-height` class (or `transitionName.height`) is not specified the change in container height will not
    be animated but instead jump to the height of the entering component instantaneously. It can, therefore, be
    omitted if all content is known to be of the same height without any adverse side-effects, and absolute positioning
    related height issues will still be avoided.

## Contributing

PRs are welcome.

## License

This software is free to use under the MIT license.
See the [LICENSE file](/LICENSE.md) for license text and copyright information.


CHANGELOG.md
### v5.0.1 (17 May 2023)

- [BREAKING] Remove deprecated findDOMNode; now requires functional child components to use `forwardRef` (#101)
- [BUGFIX] Avoid an occasional flicker when the exiting node is removed (#114)

### v4.0.5 (6 September 2021)

- [UPGRADE] Add React 17 to peer dependencies.

### v4.0.4 (30 January 2020)

- [UPGRADE] Upgrade `dom-helpers@5`.

### v4.0.3 (30 January 2020)

- [CHORE] Remove deprecated `React.createFactory`.

### v4.0.2 (11 October 2019)

- [BUGFIX] Fix child ref function called a second time with CSSTransitionGroupChild. (#61)

### v4.0.1 (11 October 2019)

- [UPGRADE] Minor and patch dependency upgrades.
- [BUGFIX] revert `dom-helpers` to `v3`. (#65,#66)

### v4.0.0 (17 September 2019)

**BROKEN RELEASE**

- [BREAKING] Rename to `UNSAFE_componentWillReceiveProps`; requires `react@>=16.3.0`.
- [UPGRADE] Remove the use of `componentWillMount` which is unsafe.
- [UPGRADE] Update dependencies; switch to Babel v7.
- [ENHANCEMENT] Add ESM build output.
- [DEPENDENCY] Remove the `warning` dependency since function refs are now standard practice.
- [CHORE] Switch to a monorepo for the lib and demo and use CRA for the demo app.
- [CHORE] Use Prettier.

### v3.0.3 (25 April 2017)

- [UPGRADE] Update dependencies to latest stable versions.
- [ENHANCEMENT] Remove the `react-transition-group` dependency. (#56)

### v3.0.2 (7 October 2017)

- [BUGFIX] Fix updating children when the current child re-renders itself. (#53)

### v2.2.2 (6 October 2017)

- [UPGRADE] Support React 16 in v2. (#55 thanks @GeKrom)

### v3.0.1 (2 October 2017)

- [BUGFIX] Prevent animating child selection with CSS rather than clearing with JS so form inputs don't lose focus on each update. (#50)
- [BUGFIX] Switch the container to relative positioning on the render before the transitions start to avoid the glitch
  in Edge without using permanent relative positioning which is not an option. (#52)
- [DOCS] Add an example where the height transition is delayed until after the leave transition. (#43)

### v3.0.0 (27 August 2017)

##### Breaking change

- [FEATURE] The `isLeaving` prop is only added to children if opted in with `notifyLeaving={true}` since it's
  a departure from `react-transition-group` features.

##### Non-breaking changes

- [ENHANCEMENT] Treat a child rendering `null` as if there is no child; specifically helps avoid errors with RR4.
- [ENHANCEMENT] Maintain component callback refs by not overwriting with string refs similar to `react-transition-group`.
- [ENHANCEMENT] Use `requestAnimationFrame` to queue the height transition rather than a timeout.
- [ENHANCEMENT] Handle the enter and leave animation of changes due to successive child updates before the current transition ends.
- [ENHANCEMENT] Clear the selection after transitions to avoid the child being selected after multiple clicks.
- [ENHANCEMENT] Entering child renders with absolute positioning since switching from relative to abs on a premature leave cancels the active enter animation.
- [ENHANCEMENT] Fix enter animation of absolutely positioned elements in Chrome not working by skipping one animation frame in the Child component.
- [ENHANCEMENT] Fix Edge glitch when render starts by always applying container `position`, `display` (use a `div`) and `overflow` styles.

### v2.2.1 (29 April 2017)

- [UPGRADE] Add a `yarn` lock file.
- [UPGRADE] Use `prop-types` instead of `React.PropTypes` for React `15.5.0`.
- [UPGRADE] Use `react-transition-group/CSSTransitionGroup` instead of `react-addons...` (thanks @ali) (#45)

### v2.2.0 (27 February 2017)

- [FEATURE] Remove element at once when transitionLeave is false. (thanks @wifilinker) (#34)
- [FEATURE] Use next component sizes when current does not exist. (thanks @wifilinker) (#34)
- [BUGFIX] Avoid silencing transition when this.state.nextChild exists. (thanks @wifilinker) (#34)
- [FEATURE] Add class with `-height-active` suffix when transitioning. (thanks @wifilinker) (#34)
- [BUGFIX] Exclude `gh-pages` from the npm package. (#37)
- [BUGFIX] Remove comma dangle after rest property. (#39)

### v2.1.0 (6 December 2016)

- [FEATURE] Add the `changeWidth` prop to have the container width animated along with the height. (thanks @wifilinker) (#31)
- [UPGRADE] Upgrade dev dependencies to latest versions; `react-bootstrap@0.30.5` fixes the demo unknown props warnings.

### v2.0.1 (14 September 2016)

- [DOCUMENTATION] Minor grammar improvement in the README.

### v2.0.0 (14 September 2016)

- [ENHANCEMENT] Add support for IE<=11 (thanks @le0nik) (#22, #23)
- [ENHANCEMENT] Provide an `isLeaving` prop on the current child component while it is leaving. (thanks @le0nik)
- [ENHANCEMENT] Add ability to provide custom `height` className (useful for `css-modules`) (thanks @le0nik) (#20, #21)
- [BUGFIX] Fix multiple remounts of children (thanks @le0nik) (#24)
- [BUGFIX] Allow the child, that is leaving, to be `null` (thanks @le0nik)

### v1.3.0 (9 July 2016)

- [ENHANCEMENT] When removing an element just added to an empty container transitions are reset to avoid getting stuck.
- [ENHANCEMENT] When starting a height transition and the entering element hasn't rendered wait another tick. (#6,#10)
- [ENHANCEMENT] Don't pass non-standard DOM props to the component as required by `react 15.2.0`. (#17)
- [UPGRADE] Use `react@15.2.0` and upgraded dev dependencies. (Should still work with `react@0.14` with `npm` warnings)
- [ENHANCEMENT] Use the spread operator rather than `object-assign`. (#16)

### v1.2.0-beta (18 April 2016)

- [UPGRADE] Add peer dependency `react@15.0.1`.
- [ENHANCEMENT] Add a `displayName` field.

### v1.1.2 (9 April 2016)

- [UPGRADE] Use the `object-assign` lib, as [React 15 does](https://github.com/facebook/react/pull/6376), rather
  than `react/lib/Object.assign` which has been removed. (#14)

### v1.1.1 (4 April 2016)

- [BUGFIX] Corrected the file name in the `main` field of package.json. (#13)
- [BUGFIX] Allow prop `transitionName` to be of type `object`. (#9,#12)

demo/src/Demo.js
import React from 'react'
import { Navbar, Nav, NavItem, Grid } from 'react-bootstrap'

const NavbarBrand = Navbar.Brand

import PageHead from './components/PageHead'
import ContentSwapper from './components/ContentSwapper'
import ContentAddRemove from './components/ContentAddRemove'
import AnimatedRouter from './components/AnimatedRouter'

import ContentLong from './components/ContentLong'
import ContentShort from './components/ContentShort'

class Demo extends React.Component {

  componentDidMount() {
    const images = [
      'img/vista1.jpg',
      'img/vista2.jpg',
      'img/vista3.jpg',
      'img/vista4.jpg',
    ]

    images.forEach(src => {
      const img = new window.Image()
      img.src = src
    })
  }

  render() {
    return (
      <div>
        <Navbar>
          <NavbarBrand>ReactCSSTransitionReplace</NavbarBrand>
          <Nav pullRight>
            <NavItem href="https://github.com/marnusw/react-css-transition-replace" target="_blank">GitHub</NavItem>
          </Nav>
        </Navbar>

        <Grid>
          <PageHead/>

          <div className="examples">
            <h2>Examples</h2>
            <p className="text-danger"><em>Click any content to trigger the transition.</em></p>

            <h3 id="cross-fade">Cross-fade transition</h3>
            <p>The <code>opacity</code> transitions for <code>enter</code> and <code>leave</code> are started
              at the same time. View the <a
                href="https://github.com/marnusw/react-css-transition-replace#cross-fading-two-components"
                target="_blank">CSS</a>.</p>

            <ContentSwapper transitionName="cross-fade" transitionEnterTimeout={1000} transitionLeaveTimeout={1000}>
              <img key="img1" src="img/vista1.jpg" width="600" height="235"/>
              <img key="img2" src="img/vista2.jpg" width="600" height="280"/>
            </ContentSwapper>

            <h3 id="fade-wait">Fade out, then fade in transition</h3>
            <p>For the first paragraph the <code>opacity</code> transition for <code>enter</code> animation is delayed
              until after the <code>leave</code> transition completes. View the <a
                href="https://github.com/marnusw/react-css-transition-replace#fade-out-then-fade-in"
                target="_blank">CSS</a>.<br/>For the second paragraph both the enter and height transitions
              are delayed.</p>

            <ContentSwapper transitionName="fade-wait" transitionEnterTimeout={1000} transitionLeaveTimeout={400}>
              <ContentLong key="long"/>
              <ContentShort key="short"/>
            </ContentSwapper>

            <br/>

            <ContentSwapper transitionName="fade-wait-delay" transitionEnterTimeout={1000} transitionLeaveTimeout={400}>
              <ContentLong key="long"/>
              <ContentShort key="short"/>
            </ContentSwapper>

            <h3 id="carousel-swap">Carousel transition</h3>
            <p>The slide animation is realised with a 2D CSS <code>transform</code>. View the <a
              href="transitions.css" target="_blank">CSS source</a>. To ensure the entering/leaving
              images are properly hidden the carousel width can be set directly on the container:
              <code>{'<ReactCSSTransitionReplace transitionName="carousel-swap" /*...*/  style={{width: 600}}>'}</code>.
            </p>

            <ContentSwapper transitionName="carousel-swap" transitionEnterTimeout={2000} transitionLeaveTimeout={2000}
                            style={{width: 600}}>
              <img key="img1" src="img/vista3.jpg" width="600" height="255"/>
              <img key="img2" src="img/vista4.jpg" width="600" height="280"/>
              <img key="img3" src="img/vista2.jpg" width="600" height="290"/>
            </ContentSwapper>

            <h3 id="roll-up">Add/Remove Content</h3>
            <p>The child component may be removed (i.e. <code>ReactCSSTransitionReplace</code> left with no children)
              which will animate the <code>height</code> going to zero along with the <code>leave</code> transition.
              A single child can subsequently be added again, triggering the inverse animation.
              View the <a href="transitions.css" target="_blank">CSS source</a> of the `roll-up` transition.
            </p>

            <ContentAddRemove transitionName="roll-up" transitionEnterTimeout={800} transitionLeaveTimeout={800}>
              <img key="img1" src="img/vista1.jpg" width="600" height="235"/>
            </ContentAddRemove>

            <h3 id="height-and-width">Height and Width animation</h3>
            <p>By setting the <code>changeWidth</code> prop the container width is managed along with the height.
              This example realizes the same effect as above using this property and just a fade CSS animation.
              In this case the <code>height</code> class should configure the transition of both the height
              and width properties: <code>transition: height .5s ease-in-out, width .5s ease-in-out;</code>
            </p>

            <ContentAddRemove
              transitionName="fade-fast" transitionEnterTimeout={500} transitionLeaveTimeout={500} changeWidth
            >
              <img key="img1" src="img/vista1.jpg" width="600" height="235"/>
            </ContentAddRemove>

            <h3 id="react-router-v4">React Router v4</h3>
            <p>Animating React-Router v4 transitions is supported. See the <a
              href="https://github.com/marnusw/react-css-transition-replace#react-router-v4"
              target="_blank">example</a> for details.</p>

            <AnimatedRouter/>
          </div>
        </Grid>
      </div>
    )
  }
}

export default Demo
